File: app/src/main/java/closer/vlllage/com/closer/pool/PoolActivity.java
package closer.vlllage.com.closer.pool;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;

import closer.vlllage.com.closer.App;
import closer.vlllage.com.closer.handler.data.PermissionHandler;
import closer.vlllage.com.closer.handler.helpers.ActivityHandler;
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler;
import closer.vlllage.com.closer.handler.helpers.CameraHandler;
import closer.vlllage.com.closer.handler.media.MediaHandler;

public abstract class PoolActivity extends FragmentActivity {
    private final Pool pool = new Pool();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        $(ApplicationHandler.class).setApp((App) getApplication());
        $(ActivityHandler.class).setActivity(this);
    }

    @Override
    protected void onDestroy() {
        pool.end();
        super.onDestroy();
    }

    protected <T extends PoolMember> T $(Class<T> member) {
        return pool.$(member);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        $(PermissionHandler.class).onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        $(CameraHandler.class).onActivityResult(requestCode, resultCode, data);
        $(MediaHandler.class).onActivityResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/pool/PoolActivity.kt
package closer.vlllage.com.closer.pool

import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.FragmentActivity

import closer.vlllage.com.closer.App
import closer.vlllage.com.closer.handler.data.AccountHandler
import closer.vlllage.com.closer.handler.data.ApiHandler
import closer.vlllage.com.closer.handler.data.PermissionHandler
import closer.vlllage.com.closer.handler.helpers.ActivityHandler
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler
import closer.vlllage.com.closer.handler.helpers.CameraHandler
import closer.vlllage.com.closer.handler.helpers.ScanQrCodeHandler
import closer.vlllage.com.closer.handler.media.MediaHandler
import com.queatz.on.On

abstract class PoolActivity : FragmentActivity() {

    val on = On()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        on<ApplicationHandler>().app = application as App
        on<ActivityHandler>().activity = this
        on<ApiHandler>().setAuthorization(on<AccountHandler>().phone)
    }

    override fun onDestroy() {
        on.off()
        super.onDestroy()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        on<PermissionHandler>().onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        on<CameraHandler>().onActivityResult(requestCode, resultCode, data)
        on<MediaHandler>().onActivityResult(requestCode, resultCode, data)
        on<ScanQrCodeHandler>().onActivityResult(requestCode, resultCode, data)
    }
}
